src/components/application_manager/src/application_manager_impl.cc

```diff
@@ -3405,6 +3408,45 @@ void ApplicationManagerImpl::OnHMILevelChanged(
     mobile_apis::HMILevel::eType from,
     mobile_apis::HMILevel::eType to) {
   LOG4CXX_AUTO_TRACE(logger_);
+  LOG4CXX_INFO(logger_, "HMI level from = " << from << " to = " << to);
+  ProcessPostponedMessages(app_id);
+  ProcessNaviApp(app_id, from, to);
+}
+
+void ApplicationManagerImpl::ProcessPostponedMessages(const uint32_t app_id) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  ApplicationSharedPtr app = application(app_id);
+  if (!app) {
+    LOG4CXX_WARN(logger_, "The app with id: " << app_id << " is not exists");
+    return;
+  }
+  MobileMessageQueue messages;
+  app->SwapMobileMessageQueue(messages);
+  std::for_each(
+      messages.begin(),
+      messages.end(),
+      [this, &app](smart_objects::SmartObjectSPtr message) {
+        const std::string functionID = MessageHelper::StringifiedFunctionID(
+            static_cast<mobile_apis::FunctionID::eType>(
+                (*message)[strings::params][strings::function_id].asUInt()));
```

Comment:
@LitvinenkoIra there can be INVALID_ENUM  that is -1. reading it as uint can give unpredictable behaviour